### PR TITLE
Chore: support gmail & yahoo smtp check by api

### PR DESCRIPTION
**Pull request by OisCircle**
## Background
Add support of smtp check by api
## What's changed
1. specific host check smtp by api
2. enable gmail & yahoo check by api

### CODE REVIEW

**Comment by OisCircle** [comment]
PR in draft status, waitting for adding tests.

**Review by OisCircle — COMMENTED** [review]
reviewed

File: smtp_by_api_yahoo.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+	for _, item := range res.Errors {
+		if item.Name == "userId" && item.Error == "IDENTIFIER_EXISTS" {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (y yahoo) sendValidateRequest(domain, username, acrumb, sessionIndex string, cookies []*http.Cookie) ([]byte, error) {
+	data, err := json.Marshal(struct {
```
Comment (lryong) [inline smtp_by_api_yahoo.go:100]: Define a separate struct, instead of putting it in a function？

File: smtp.go [hunk 2]
```diff
@@ -33,26 +34,35 @@ func (v *Verifier) CheckSMTP(domain, username string) (*SMTP, error) {
 	}
 
 	var ret SMTP
+	var err error
+	email := fmt.Sprintf("%s@%s", username, domain)
 
 	// Dial any SMTP server that will accept a connection
-	client, err := newSMTPClient(domain, v.proxyURI)
+	client, mx, err := newSMTPClient(domain, v.proxyURI)
```
Comment (lryong) [inline smtp.go:41]: Is it necessary to check the domain obtained by mx resolution, instead of directly determining the domain name

Comment (OisCircle) [reply smtp.go:41]: Take an example, `xxx@aftership.com` belongs to gmail and its mx.Host is `aspmx.l.google.com.`. If we determine by domain, `@aftership.com` will not be recognized as gmail.

File: smtp_by_api_yahoo.go (hunk 1 above)
Comment (OisCircle) [reply smtp_by_api_yahoo.go:110]: ok

**Review by lryong — APPROVED** [review]

File: verifier.go [hunk 3]
```diff
@@ -131,6 +133,30 @@ func (v *Verifier) EnableSMTPCheck() *Verifier {
...
+}
+
+func (v *Verifier) DisableGmailCheckByAPI() {
+	delete(v.apiVerifiers, "gmail")
+}
+
+func (v *Verifier) DisableYahooCheckByAPI() {
+	delete(v.apiVerifiers, "yahoo")
+}
+
```
Comment (git-hulk) [inline verifier.go:159]: It's a bit wordy to create a functioning pair for every vendor. Can change to something like `Enable/DisableAPIVerifer(name string)`.

**Comment by git-hulk** [comment]
@OisCircle @lryong I'm not familiar with the email vendor, so not sure if it has any risks in using the signup process to check if the user exists. If yes, I think we should add some comments to notice this. Others are good to me and left one comment inline.

**Comment by lryong** [comment]
Since it is not practical to check the existence of email addresses through smtp protocol, we have researched that large free email providers such as gmail and yahoo can check email addresses through relevant API interface methods, so we try to add the ability to check the validity of email addresses here.
For more information, please refer to [check-if-email-exists/blob/master/core/src/smtp/yahoo.rs](https://github.com/reacherhq/check-if-email-exists/blob/master/core/src/smtp/yahoo.rs)

@git-hulk @NeoCN

**Comment by git-hulk** [comment]
> Since it is not practical to check the existence of email addresses through smtp protocol, we have researched that large free email providers such as gmail and yahoo can check email addresses through relevant API interface methods, so we try to add the ability to check the validity of email addresses here. For more information, please refer to [check-if-email-exists/blob/master/core/src/smtp/yahoo.rs](https://github.com/reacherhq/check-if-email-exists/blob/master/core/src/smtp/yahoo.rs)
> 
> @git-hulk @NeoCN

Yes, it sounds good. Need to take care of the risks if plan to use it in production.

File: verifier.go (hunk 3 above)
Comment (OisCircle) [reply verifier.go:159]: Hi. @git-hulk 
I have left a comment before.
> // Certain API verifiers require additional input, for instance, Hotmail necessitates the use of a WebDriver.

If using the way you mentioned above, the function parameter will expand as time goes on. Thus i thought every APIVerifier should have it's own custom initialize function and params.

Please refer to https://github.com/reacherhq/check-if-email-exists/blob/master/core/src/smtp/microsoft/hotmail.rs#LL57C2-L57C11

I think we will extend new verifiers later.

**Comment by OisCircle** [comment]
> @OisCircle @lryong I'm not familiar with the email vendor, so not sure if it has any risks in using the signup process to check if the user exists. If yes, I think we should add some comments to notice this. Others are good to me and left one comment inline.

**Comment by OisCircle** [comment]
> Since it is not practical to check the existence of email addresses through smtp protocol, we have researched that large free email providers such as gmail and yahoo can check email addresses through relevant API interface methods, so we try to add the ability to check the validity of email addresses here. For more information, please refer to [check-if-email-exists/blob/master/core/src/smtp/yahoo.rs](https://github.com/reacherhq/check-if-email-exists/blob/master/core/src/smtp/yahoo.rs)
> 
> @git-hulk @NeoCN

I have added comments to inform the user about the potential risks and related knowledge on abusing.

File: verifier.go (hunk 3 above)
Comment (OisCircle) [reply verifier.go]: Got it, thanks.

**Review by git-hulk — APPROVED** [review]
